### PR TITLE
#3426 Show a user message in case docation delete/ void is used.

### PR DESCRIPTION
**Pull request by dragospodariu96**
https://github.com/metasfresh/metasfresh/issues/6100

### CODE REVIEW

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java [hunk 1]
```diff
@@ -98,38 +68,55 @@
...
 	private final MQuery zoomQuery;
 
-	/** Direct Access Query (i.e. SELECT Key, Value, Name ... FROM TableName WHERE KeyColumn=?) */
+	/**
+	 * Direct Access Query (i.e. SELECT Key, Value, Name ... FROM TableName WHERE KeyColumn=?)
+	 */
 	private TranslatableParameterizedString sqlQueryDirect = TranslatableParameterizedString.EMPTY;
-	/** Parent Flag */
+	/**
+	 * Parent Flag
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java]: useless javadoc

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java [hunk 2]
```diff
@@ -98,38 +68,55 @@
...
+	 */
 	private boolean IsParent = false;
-	/** Key Flag */
+	/**
+	 * Key Flag
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java]: useless javadoc

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java [hunk 3]
```diff
@@ -98,38 +68,55 @@
...
+	 */
 	private boolean IsKey = false;
 	private IValidationRule _validationRule = NullValidationRule.instance;
 	private IValidationRule _validationRuleEffective = null; // lazy
 
-	/** WindowNo */
+	/**
+	 * WindowNo
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java]: useless javadoc

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java [hunk 4]
```diff
@@ -98,38 +68,55 @@
...
+	 */
 	private int WindowNo;
 
-	/** AD_Reference_ID */
+	/**
+	 * AD_Reference_ID
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/model/MLookupInfo.java]: alright, this one has added value 👍

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/ValueNamePair.java [hunk 5]
```diff
@@ -35,6 +36,12 @@
 {
 	private static final long serialVersionUID = -8315081335749462163L;
 
+	public static final ValueNamePair EMPTY = new ValueNamePair("", "", null/* help */);
+
+	/**
+	 * The Validation Message
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/compiere/util/ValueNamePair.java]: useless javadoc

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5550520_sys_gh3426_AddedValidationMessages.sql [hunk 6]
```diff
@@ -0,0 +1,226 @@
+-- 2020-01-28T16:24:30.008Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+INSERT INTO AD_Element (AD_Client_ID, AD_Element_ID, AD_Org_ID, Created, CreatedBy, EntityType, IsActive, Name, PrintName, Updated, UpdatedBy)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5550520_sys_gh3426_AddedValidationMessages.sql]: please add a useful description

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java [hunk 7]
```diff
@@ -808,6 +810,17 @@ protected NamePair fetch(final ResultSet rs) throws SQLException
 			final String name = getDisplayName(rs, isActive);
 			final String description = rs.getString(MLookupFactory.COLUMNINDEX_Description);
 
+			String validationMessage;
+			try
+			{
+				validationMessage = rs.getString(MLookupFactory.COLUMNINDEX_ValidationMsg);
+			}
+			catch (Exception e)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java]: i looked only superficially, but I can't see why we need to take precautions here. 
* why not make the SQL such that the `rs` always contains the validation message?
* if that's not practical, then add a flag or something to tell the iteration if it shall try to get the validationMessage

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java [hunk 8]
```diff
@@ -808,6 +810,17 @@ protected NamePair fetch(final ResultSet rs) throws SQLException
...
+			{
+				validationMessage = null;
+				logger.info("Only used in DocAction dropdown");
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java]: whether the value is set only in DocAction dropdown should not be this code's concern

**Comment by metas-ts** [comment]
@dragospodariu96 despite my change-requests, AFAIU you can deploy the current code to a dev-instance and notify @petrican. That way he could already start working on the related frontend issue.

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java (hunk 7 above)
Comment (teosarca) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java]: agree

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java (hunk 8 above)
Comment (teosarca) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/ad/service/impl/LookupDAO.java]: exactly.
more, avoid `log.info` for repetitive stuff. use log.debug/trace

**Comment by metas-ts** [comment]
Yesterday i only saw individual changes in the code
Having slept over it, i think we shall not go into this direction at all. 
* there are more than 2k `AD_Ref_List` records. we can't add another column that is used in four of them tops and does nothing in the others
* the mixture of text&syntax won't work out like this. i can already see people changing it from "question||yes||no" ("frage||ja||nein") to "question||no||yes" or "question||yes||no||cancel" and expecting it to work in some ways.
  * yesterday in the review i asked for an explanative help-text for the new field. just imagine how big that text would be so make sure that future users understand how the field shall be used.

Note: i know think that `JSONDocumentAction` should contain something like an optional JSONConfirmaRequestDescriptor (to be created) that can tell the frontend to request the user to confirm an action before proceeding

**Pull request closed without merging**